### PR TITLE
Feature: GMSD and MultiScale GMSD

### LINKED ISSUE #67: Metric: Gradient Magnitude Similarity and it's multi-scale version
**Is your feature request related to a problem? Please describe.**
New metrics to measure full-reference image quality based on gradients similarity. 
It was shown to improve Super-Resolution quality compared to usage of only MSE loss. 

**Describe the solution you'd like**
Add Gradient magnitude similarity deviation (GMSD) https://arxiv.org/pdf/1308.3052.pdf
And Multiscale GMSD https://ieeexplore.ieee.org/document/7952357

### PR DESCRIPTION
**Pull request by zakajd**
Closes #67

### CODE REVIEW

File: photosynthesis_metrics/gmsd.py [hunk 1]
```diff
@@ -0,0 +1,253 @@
...
+
+    # Compute GMSD
+    gmsd = torch.pow(gms - mean_gms, 2).mean(dim=[1, 2, 3], keepdims=True).sqrt()
+    
+    if reduction == 'mean':
+        return gmsd.mean()
+    elif reduction == 'sum':
+        return gmsd.sum()
+    else:
+        return gmsd
```
Comment (denproc) [inline photosynthesis_metrics/gmsd.py:84]: - Let's specify the dimension for the `mean` and `sum` operations. Therefore, we will have passive check that everything is calculated right and improve readability that we perform operation for batch dimension.
- Let's add check for specified `reduction` modes.
```suggestion
    if reduction == 'mean':
        return gmsd.mean(dim=0)
    elif reduction == 'sum':
        return gmsd.sum(dim=0)
    elif reduction != 'none':
        raise ValueError(f'Expected reduction modes are "mean"|"sum"|"none", got {reduction}')
    return gmsd
```

File: photosynthesis_metrics/gmsd.py [hunk 2]
```diff
@@ -0,0 +1,253 @@
...
+        r"""Computation of Gradient Magnitude Similarity Deviation (GMSD) as a loss function.
+
+        Args:
+            prediction: Tensor of prediction of the network.
+            target: Reference tensor.
+
+        Returns:
+            Value of GMSD loss to be minimized. 0 <= GMSD loss <= 1.
+        """
+        _validate_input(x=prediction, y=target, scale_weights=None)
```
Comment (denproc) [inline photosynthesis_metrics/gmsd.py:205]:
```suggestion
        _validate_input(x=prediction, y=target, scale_weights=self.scale_weights)
```

File: photosynthesis_metrics/gmsd.py [hunk 3]
```diff
@@ -0,0 +1,253 @@
...
+        scale_weights = self.scale_weights.to(prediction)
+        ms_gmds = []
+        for scale in range(4):
+            if scale > 0:
+                # Average by 2x2 filter and downsample
+                padding = (prediction.shape[2] % 2, prediction.shape[3] % 2)
+                prediction = F.avg_pool2d(prediction, kernel_size=2, padding=padding)
+                target = F.avg_pool2d(target, kernel_size=2, padding=padding)
+        
+            score = _gmsd(prediction, target, reduction=None)
```
Comment (denproc) [inline photosynthesis_metrics/gmsd.py:224]:
```suggestion
            score = _gmsd(prediction, target, reduction='none')
```

File: photosynthesis_metrics/gmsd.py [hunk 4]
```diff
@@ -0,0 +1,253 @@
...
+            ms_gmds.append(score)
+        
+        # Concat results in different scales and multiply by weight
+        ms_gmds_val = scale_weights.view(1, 4, 1, 1) * (torch.cat(ms_gmds, dim=1) ** 2)
+        # Sum and take sqrt per-image
+        ms_gmds_val = torch.sqrt(torch.sum(ms_gmds_val, dim=-1))
```
Comment (denproc) [inline photosynthesis_metrics/gmsd.py:230]: The sum is performed for `dim=4`, which equals 1. Shouldn't the operation be performed for `dim=(1,2,3)`?
```suggestion
        ms_gmds_val = torch.sqrt(torch.sum(ms_gmds_val, dim=(1, 2, 3)))
```

File: photosynthesis_metrics/gmsd.py [hunk 5]
```diff
@@ -0,0 +1,253 @@
...
+            gamma = 2 / (1 + self.beta2 * torch.exp(-self.beta3 * ms_gmds_val)) - 1
+            
+            score = gamma * ms_gmds_val + (1 - gamma) * self.beta1 * rmse_chrome
+            
+        if self.reduction == 'mean':
+            score = torch.mean(score)
+        elif self.reduction == 'sum':
+            score = torch.sum(score)
+
+        return score
```
Comment (denproc) [inline photosynthesis_metrics/gmsd.py:253]:
```suggestion
        if self.reduction == 'mean':
            score = torch.mean(score, dim=0)
        elif self.reduction == 'sum':
            score = torch.sum(score, dim=0)
        elif self.reduction != 'none':
            raise ValueError(f'Expected reduction modes are "mean"|"sum"|"none", got {self.reduction}')

        return score
```

File: photosynthesis_metrics/gmsd.py [hunk 6]
```diff
@@ -0,0 +1,253 @@
...
+    def compute_metric(self, prediction: torch.Tensor, target: torch.Tensor) -> torch.Tensor:
+        
+        if self.data_range == 255:
+            prediction = prediction / 255.
+            target = target / 255.
+
+        # Average by 2x2 filter and downsample
+        padding = (prediction.shape[2] % 2, prediction.shape[3] % 2)
+        prediction = F.avg_pool2d(prediction, kernel_size=2, stride=2, padding=padding)
+        target = F.avg_pool2d(target, kernel_size=2, stride=2, padding=padding)
```
Comment (denproc) [inline photosynthesis_metrics/gmsd.py:139]: According to [last paper](http://www.cse.ust.hk/~psander/docs/gradsim.pdf) there is no need in downsample for GMSD. 
Also, it is stated that first GMDS at multiscale approach is calculated for original sized images. In current implementation, we do downsampling first, then calculate GMDS.

File: photosynthesis_metrics/gmsd.py [hunk 7]
```diff
@@ -0,0 +1,253 @@
...
+            
+        scale_weights = self.scale_weights.to(prediction)
+        ms_gmds = []
+        for scale in range(4):
+            if scale > 0:
+                # Average by 2x2 filter and downsample
+                padding = (prediction.shape[2] % 2, prediction.shape[3] % 2)
+                prediction = F.avg_pool2d(prediction, kernel_size=2, padding=padding)
+                target = F.avg_pool2d(target, kernel_size=2, padding=padding)
+
```
Comment (denproc) [inline photosynthesis_metrics/gmsd.py:223]: If there is no need in avgpooling at first stage, let's put `_gmsd` call first then pooling as it is done at MS-SSIM.

File: photosynthesis_metrics/gmsd.py [hunk 8]
```diff
@@ -0,0 +1,253 @@
...
+        
+        score = ms_gmds_val
+        
+        if self.chromatic:
+            assert prediction.size(1) == 3, "Chromatic component can be computed only for RGB images!"
+            
+            # Convert to YIQ color space https://en.wikipedia.org/wiki/YIQ
+            iq_weights = torch.tensor([[0.5959, -0.2746, -0.3213], [0.2115, -0.5227, 0.3112]]).t()
+            prediction_iq = torch.matmul(prediction.permute(0, 2, 3, 1), iq_weights).permute(0, 3, 1, 2)
+            target_iq = torch.matmul(target.permute(0, 2, 3, 1), iq_weights).permute(0, 3, 1, 2)
```
Comment (denproc) [inline photosynthesis_metrics/gmsd.py:240]: The prediction and target are not equal to input ones as avgpooling was performed several times.

File: photosynthesis_metrics/gmsd.py [hunk 9]
```diff
@@ -0,0 +1,253 @@
+"""
```
Comment (denproc) [inline photosynthesis_metrics/gmsd.py:1]: Only classes were defined here. Should we define the methods for GMSD and MS-GMSD as we did for SSIM and MS-SSIM?

File: photosynthesis_metrics/gmsd.py (hunk 6 above)
Comment (zakajd) [reply photosynthesis_metrics/gmsd.py:139]: See original paper for [GMSD]( https://arxiv.org/pdf/1308.3052.pdf) Section 3.B 
>In addition, as in the implementations of SSIM [8] and FSIM [7], the images r and d are first filtered by a 2×2 average filter, and then down-sampled by a factor of 2

They also do it in their MATLAB code. Another reason for this is weights for MS-GMSD, where the biggest weight in on scale 2, which is exactly blurred and downsampled image.

File: photosynthesis_metrics/gmsd.py (hunk 8 above)
Comment (zakajd) [reply photosynthesis_metrics/gmsd.py:240]: See MS-GMSD paper, chapter 3.3
> It should be noted that chrominance dissimilarity is only evaluated at Scale M because the spatial resolution to color stimulation is much lower than luminance for vision perception.

File: photosynthesis_metrics/gmsd.py (hunk 9 above)
Comment (zakajd) [reply photosynthesis_metrics/gmsd.py:1]: Why? The reason to have separate class for `SSIM` is that when it's used as a measure it's value _maximised_ and when used as a loss it's value _minimised_. Here for both cases value of GMSD is minimised, so class `GMSDLoss` can be used as a measure without any additional changes.

Comment (denproc) [reply photosynthesis_metrics/gmsd.py:1]: Well, actually no. As I understood, the solution was to return true value and describe it in the documentation that user should take it into account.

File: photosynthesis_metrics/gmsd.py (hunk 2 above)
Comment (zakajd) [reply photosynthesis_metrics/gmsd.py]: We have different number of weights in MS-SSIM and MS-GMSD (5 and 4 respectively). So this change leads to assertion errors in `_validate_input` function.

Comment (denproc) [reply photosynthesis_metrics/gmsd.py]: `_validate_input` will support flexible number of weights after #55. It's already implemented.

Comment (denproc) [reply photosynthesis_metrics/gmsd.py]: May user use custom weights of different size? like `[0.4, 0.5, 0.3, 0.2, 0.1]`?

File: tests/test_gmsd.py [hunk 10]
```diff
@@ -0,0 +1,111 @@
...
+    except Exception as e:
+        pytest.fail(f"Unexpected error occurred: {e}")
+
+
+def test_multi_scale_gmsd_small_for_tensors_with_contrast_difference():
+    loss = MultiScaleGMSDLoss(chromatic=True)
+    target = torch.ones(3, 3, 256, 256)
+    prediction = torch.zeros(3, 3, 256, 256)
+    measure = loss(prediction, target)
+    assert measure <= 1e-6, \
```
Comment (denproc) [inline tests/test_gmsd.py:110]:
```suggestion
    assert measure.abs() <= 1e-6, \
```

File: tests/test_gmsd.py [hunk 11]
```diff
@@ -0,0 +1,111 @@
...
+        MultiScaleGMSDLoss()
+    except Exception as e:
+        pytest.fail(f"Unexpected error occurred: {e}")
+
+
+def test_multi_scale_gmsd_zero_for_equal_tensors(prediction: torch.Tensor):
+    loss = MultiScaleGMSDLoss()
+    target = prediction.clone()
+    measure = loss(prediction, target)
+    assert measure <= 1e-6, f'MultiScaleGMSD for equal tensors must be 0, got {measure}'
```
Comment (denproc) [inline tests/test_gmsd.py:72]:
```suggestion
    assert measure.abs() <= 1e-6, f'MultiScaleGMSD for equal tensors must be 0, got {measure}'
```

File: tests/test_gmsd.py [hunk 12]
```diff
@@ -0,0 +1,111 @@
...
+        GMSDLoss()
+    except Exception as e:
+        pytest.fail(f"Unexpected error occurred: {e}")
+
+
+def test_gmsd_zero_for_equal_tensors(prediction: torch.Tensor):
+    loss = GMSDLoss()
+    target = prediction.clone()
+    measure = loss(prediction, target)
+    assert measure <= 1e-6, f'GMSD for equal tensors must be 0, got {measure}'
```
Comment (denproc) [inline tests/test_gmsd.py:29]:
```suggestion
    assert measure.abs() <= 1e-6, f'GMSD for equal tensors must be 0, got {measure}'
```

File: photosynthesis_metrics/gmsd.py [hunk 13]
```diff
@@ -0,0 +1,256 @@
...
+        _validate_input(x=prediction, y=target, scale_weights=self.scale_weights)
+        prediction, target = _adjust_dimensions(x=prediction, y=target)
+
+        return self.compute_metric(prediction, target)
+    
+    def compute_metric(self, prediction: torch.Tensor, target: torch.Tensor) -> torch.Tensor:
+        if self.data_range == 255:
+            prediction = prediction / 255.
+            target = target / 255.
+
```
Comment (denproc) [inline photosynthesis_metrics/gmsd.py:215]: Following #35, there should be an Assertion on the image sizes.

File: photosynthesis_metrics/gmsd.py (hunk 2 above)
Comment (zakajd) [reply photosynthesis_metrics/gmsd.py]: Weights should sum to 1. A added weights normalization in `__init__`, so even when user writes `[2., 4., 1., 1.]` everything should work fine.

File: photosynthesis_metrics/gmsd.py (hunk 13 above)
Comment (zakajd) [reply photosynthesis_metrics/gmsd.py:216]: Difference here is that for MS-SSIM minimal size is 32x32 and lot's of small datasets do not satisfy this condition. While for MS-GMSD minimal image size is 16x16. I doubt anyone will ever try to compute this metric between such small patches.

File: photosynthesis_metrics/gmsd.py (hunk 2 above)
Comment (denproc) [reply photosynthesis_metrics/gmsd.py]: The point is to use 5 weights instead of 4. Let's make them [0.2, 0.2, 0.2, 0.2, 0.2] or [0.3, 0.3, 0.3].

File: photosynthesis_metrics/gmsd.py (hunk 13 above)
Comment (denproc) [reply photosynthesis_metrics/gmsd.py:216]: The size depends on the number of scales used to calculate the measure. If user would like to use [.1]*10 weights, the size should be at least 2**len(weights) + 1.

**Comment by zakajd** [comment]
Discussed with @denproc that it's better to have support for custom number of scales, than not to have. 
Now user can write `scale_weights = [1., 5., 2.]` and everything will work.
